File: worker/lib/faction/schema/organized_crime.ex
# Copyright (C) 2021-2025 tiksan
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

defmodule Tornium.Schema.OrganizedCrime do
  @moduledoc """
  Organized crime.

  NOTE: Update __MODULE__.map/1 when more fields are added to this schema.
  """

  # TODO: Add moduledoc

  alias Tornium.Repo
  use Ecto.Schema
  import Ecto.Query

  @type t :: %__MODULE__{
          oc_id: integer(),
          oc_name: String.t(),
          oc_difficulty: integer(),
          faction: Tornium.Schema.Faction.t(),
          status: :recruiting | :planning | :successful | :failure | :expired,
          created_at: DateTime.t(),
          planning_started_at: DateTime.t() | nil,
          ready_at: DateTime.t() | nil,
          expires_at: DateTime.t(),
          executed_at: DateTime.t() | nil,
          slots: [Tornium.Schema.OrganizedCrimeSlot.t()],
          assigned_team_id: Ecto.UUID.t(),
          assigned_team: Tornium.Schema.OrganizedCrimeTeam.t(),
          assigned_team_at: DateTime.t()
        }

  @primary_key {:oc_id, :integer, autogenerate: false}
  schema "organized_crime_new" do
    field(:oc_name, :string)
    field(:oc_difficulty, :integer)

    belongs_to(:faction, Tornium.Schema.Faction, references: :tid)

    field(:status, Ecto.Enum, values: [:recruiting, :planning, :successful, :failure, :expired])

    field(:created_at, :utc_datetime)
    field(:planning_started_at, :utc_datetime)
    field(:ready_at, :utc_datetime)
    field(:expires_at, :utc_datetime)
    field(:executed_at, :utc_datetime)

    has_many(:slots, Tornium.Schema.OrganizedCrimeSlot, foreign_key: :oc_id)

    # Tornium-specific data
    belongs_to(:assigned_team, Tornium.Schema.OrganizedCrimeTeam, references: :guid, type: Ecto.UUID)
    field(:assigned_team_at, :utc_datetime)
  end

  # TODO: Add documentation
  @spec upsert_all(entries :: [t()]) :: [t()]
  def upsert_all(entries) when is_list(entries) do
    # TODO: Check the type of the head of the list
    {_, returned_entries} =
      entries
      |> map()
      |> (&Repo.insert_all(Tornium.Schema.OrganizedCrime, &1,
            on_conflict: {:replace_all_except, [:oc_id, :assigned_team_id, :assigned_team_at]},
            conflict_target: [:oc_id],
            returning: true
          )).()

    slot_users =
      entries
      |> Enum.flat_map(fn %Tornium.Schema.OrganizedCrime{slots: slots} -> slots end)
      |> Enum.map(fn %Tornium.Schema.OrganizedCrimeSlot{user_id: user_id} -> user_id end)
      |> Enum.reject(&is_nil/1)
      |> Enum.uniq()
      |> Enum.map(
        &%{
          tid: &1
        }
      )

    Repo.insert_all(Tornium.Schema.User, slot_users, on_conflict: :nothing, conflict_target: [:tid])

    returned_slot_entries =
      entries
      |> Enum.flat_map(fn %Tornium.Schema.OrganizedCrime{slots: slots} ->
        slots
      end)
      |> Tornium.Schema.OrganizedCrimeSlot.upsert_all()
      |> Enum.group_by(& &1.oc_id)

    Enum.map(returned_entries, fn %Tornium.Schema.OrganizedCrime{oc_id: oc_id} = oc ->
      # TODO: Determine a better way to implement this
      # Maybe use something similar to https://hexdocs.pm/ecto/Ecto.Repo.html#c:load/2
      Map.replace(oc, :slots, returned_slot_entries[oc_id])
    end)
  end

  @spec map(entries :: [t()]) :: list(map())
  def map(entries) when is_list(entries) do
    Enum.map(
      entries,
      fn %Tornium.Schema.OrganizedCrime{
           oc_id: oc_id,
           oc_name: oc_name,
           oc_difficulty: oc_difficulty,
           faction_id: faction_id,
           status: status,
           created_at: created_at,
           planning_started_at: planning_started_at,
           ready_at: ready_at,
           expires_at: expires_at,
           executed_at: executed_at,
           assigned_team_id: assigned_team_id,
           assigned_team_at: assigned_team_at
         } ->
        %{
          oc_id: oc_id,
          oc_name: oc_name,
          oc_difficulty: oc_difficulty,
          faction_id: faction_id,
          status: status,
          created_at: created_at,
          planning_started_at: planning_started_at,
          ready_at: ready_at,
          expires_at: expires_at,
          executed_at: executed_at,
          assigned_team_id: assigned_team_id,
          assigned_team_at: assigned_team_at
        }
      end
    )
  end

  @doc """
  Assign an organized crime team to an organized crime in the DB.

  The `assigned_team_at` timestamp will be set to the current time indicating when the OC team was assigned to the
  OC.
  """
  @spec update_assigned_team(team :: Tornium.Schema.OrganizedCrimeTeam.t(), crime :: t()) :: {non_neg_integer(), nil}
  def update_assigned_team(
        %Tornium.Schema.OrganizedCrimeTeam{guid: team_guid} = _team,
        %__MODULE__{oc_id: oc_id} = _crime
      ) do
    __MODULE__
    |> update([c], set: [assigned_team_id: ^team_guid, assigned_team_at: ^DateTime.utc_now()])
    |> where([c], c.oc_id == ^oc_id)
    |> Repo.update_all([])
  end
end

File: commons/tornium_commons/models/organized_crime.py
# Copyright (C) 2021-2025 tiksan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations

import typing
from functools import lru_cache

from peewee import (
    BigIntegerField,
    CharField,
    DateTimeField,
    DeferredForeignKey,
    ForeignKeyField,
    SmallIntegerField,
)
from tornium_oc_graph import calculate_ev, calculate_probability

from .base_model import BaseModel
from .faction import Faction

if typing.TYPE_CHECKING:
    from .organized_crime_slot import OrganizedCrimeSlot


class OrganizedCrime(BaseModel):
    class Meta:
        table_name = "organized_crime"

    oc_id = BigIntegerField(primary_key=True)
    oc_name = CharField(null=False)
    oc_difficulty = SmallIntegerField(null=False)

    faction = ForeignKeyField(Faction, null=False)

    status = CharField(null=False)

    created_at = DateTimeField(null=False)
    planning_started_at = DateTimeField(default=None, null=True)
    ready_at = DateTimeField(default=None, null=True)
    expires_at = DateTimeField(default=None, null=True)
    executed_at = DateTimeField(default=None, null=True)

    # Tornium-specific OC data
    assigned_team = DeferredForeignKey("OrganizedCrimeTeam", default=None, null=True)

    @classmethod
    @lru_cache
    def oc_names(cls) -> typing.List[str]:
        return [crime.oc_name for crime in OrganizedCrime.select().distinct(OrganizedCrime.oc_name)]

    def to_dict(self) -> dict:
        # Skip the `assigned_team` to avoid circular imports
        # At this time, there isn't a purpose in returning that, but if necessary in the future, this can
        # be done with an optional parameter.

        return {
            "oc_id": self.oc_id,
            "oc_name": self.oc_name,
            "oc_difficulty": self.oc_difficulty,
            "faction_id": self.faction_id,
            "status": self.status,
            "created_at": self.created_at.timestamp(),
            "planning_started_at": None if self.planning_started_at is None else self.planning_started_at.timestamp(),
            "ready_at": None if self.ready_at is None else self.ready_at.timestamp(),
            "expires_at": self.expires_at.timestamp(),
            "executed_at": None if self.executed_at is None else self.executed_at.timestamp(),
        }

    @staticmethod
    def expected_value(oc_name: str, slots: typing.List[OrganizedCrimeSlot], default=None) -> float:
        succcess_map = {}

        slot: OrganizedCrimeSlot
        for slot in slots:
            if slot.user_id is None and default is None:
                raise ValueError(f"{slot.crime_position} #{slot.crime_position_index} is not filled")

            position = f"{'_'.join(slot.crime_position.lower().split(' '))}_{slot.crime_position_index}"
            succcess_map[position] = default if slot.user_id is None else slot.user_success_chance / 100

        return calculate_ev(oc_name, succcess_map)

    @staticmethod
    def probability(oc_name: str, slots: typing.List[OrganizedCrimeSlot], default=None) -> float:
        succcess_map = {}

        slot: OrganizedCrimeSlot
        for slot in slots:
            if slot.user_id is None and default is None:
                raise ValueError(f"{slot.crime_position} #{slot.crime_position_index} is not filled")

            position = f"{"_".join(slot.crime_position.lower().split(" "))}_{slot.crime_position_index}"
            succcess_map[position] = default if slot.user_id is None else slot.user_success_chance / 100

        return calculate_probability(oc_name, succcess_map)
